### ISSUE TITLE
Failure to report short paths with intervening packages

### ISSUE THREAD

**Issue by goldfirere** [ocaml/merlin#1643]
Suppose I have

```ocaml
open Core

let _ = Filename.dirname
```

and ask for the type of the `_` (with `-short-paths`). I get `Base.string -> Base.string`, instead of the desired `string -> string`. This seems like a bug.

My testing setup: a dune project, with this dune file:

```
(library
 (public_name merlin_bug)
 (name merlin_bug)
 (libraries core))
```

After building, this invocation shows the `Base.string -> Base.string` type:

```
ocamlmerlin single type-enclosing -position 3:4 -index 0 -filename merlin_bug.ml -verbosity 0 -short-paths < merlin_bug.ml
```

**Comment by voodoos** [comment]
I don't think this is an issue with `-short-paths`. The type of `Core.Filename.dirname` really is `Base.string -> Base.string`: https://ocaml.org/p/core/v0.16.0/doc/Core/Filename/index.html#val-dirname

Since `Base` is not opened there is no reason for short path to remove these prefixes.

**Comment by goldfirere** [comment]
But `Base.string` is an abbreviation for `string`. Is that not what `-short-paths` is trying to address?

**Comment by goldfirere** [comment]
Ah -- I didn't realize that.

I edited the file to be

```
open Core

module Foo = Base.Int

let _ = Filename.dirname
```

which would seem to force loading of `Base`. But I still see `Base.string -> Base.string` in the output (after adjusting the line number).

**Comment by voodoos** [comment]
In my attempt I was able to get the short paths by doing: 
```
open Base
open Core

let _ = Filename.dirname
```

**Comment by goldfirere** [comment]
That doesn't surprise me. But I would expect not to need to `open Base` to get the short paths -- that is, unless I'm misunderstanding the goal of `-short-paths`.

**Comment by goldfirere** [comment]
To expand a bit: My understanding of `-short-paths` is that it instructs OCaml / merlin to produce the shortest path it knows that is equal to a given path. `string` is definitely shorter than `Base.string`. And, regardless of what is opened, the compiler has access to the knowledge that `Base.string = string`. So, perhaps I have my picture of `-short-paths` wrong, or maybe my picture of what the compiler knows is wrong. But my last example above, which uses `Base` directly, seems to suggest that even if `Base` is all loaded (and thus that the compiler really does know that `Base.string = string`), it still doesn't produce the shortest path.

**Comment by yminsky** [comment]
Yeah, I'm pretty confused by this one too. Base.string and string are both available. Why should I need an open to get string to be chosen? string comes in from Pervasives, right? What am I missing here?

**Comment by voodoos** [comment]
As discussed offline, this indeed looks like a bug. There is a long overdue overhaul of `-short-path` that we hope to tackle with an eventual upstreaming in the future... Meanwhile I will investigate, it might be a low-hanging fruit. 

@lpw25 do you know if this behavior is a regression or something that never was handled by the feature ?

**Comment by lpw25** [comment]
This isn't really a bug, it is behaving according to the spec. I think that type-checking that file does not require loading the `.cmi` files that contain the information that `Base.string` is equal to `string`. Short-paths is currently forbidden from loading any additional `.cmi` files, so it cannot shorten `Base.string` to `string`.

I think it would be worth changing the spec to be better behaved, either by completely redesigning the whole idea of short paths, or by adjusting the rules around `.cmi` files, but I expect either of those to be a tricky to get right. Beyond that I'm not sure that there is much that can be done.
